### PR TITLE
GH-37673: [C++][FlightRPC] Improve error message if binding port fails

### LINKED ISSUE #37673: [C++][Flight] Improve error message when port cannot be bound with gRPC
### Describe the enhancement requested

Currently, when starting a flight server fails because the call to `arrow::flight::FlightServerBase::Init()` fails internally at the gRPC `BuildAndStart()` step, the error message is rather vague about the reason.

A pretty common error while starting the server is that the supplied port cannot be used; this is indicated by `selected_port` that is populated after `BuildAndStart()`, so the error message could be improved here.

### Component(s)

FlightRPC

### PR DESCRIPTION
**Pull request by romankarlstetter**
See https://github.com/apache/arrow/issues/37673 for more info
* Closes: #37673

### CODE REVIEW

File: cpp/src/arrow/flight/transport/grpc/grpc_server.cc [hunk 1]
```diff
@@ -616,6 +616,10 @@ class GrpcServerTransport : public internal::ServerTransport {
 
     grpc_server_ = builder.BuildAndStart();
     if (!grpc_server_) {
+      if (port == 0) {
+        return Status::Invalid("Server did not start properly. Unable to bind to port ",
```
Comment (lidavidm) [inline cpp/src/arrow/flight/transport/grpc/grpc_server.cc:620]: I think IOError makes more sense?

**Review by lidavidm — COMMENTED** [review]
Thank you, though fundamentally gRPC just doesn't give us anything useful here (instead opting to print to stderr)

File: cpp/src/arrow/flight/transport/grpc/grpc_server.cc (hunk 1 above)
Comment (romankarlstetter) [reply cpp/src/arrow/flight/transport/grpc/grpc_server.cc:620]: `IOError` is also fine for me

**Comment by romankarlstetter** [comment]
Sorry, renaming the master branch in my local fork was a bad idea, I guess... recreated the pull request (with the suggested changed) here: #37689

**Pull request closed without merging**